File: src/io/innofang/singleton/example/java/ThreadSafeStaticInnerClassSingleton.java
package io.innofang.singleton.example.java;

/**
 * Created by Inno Fang on 2017/8/12.
 */
public class ThreadSafeStaticInnerClassSingleton {

    private ThreadSafeStaticInnerClassSingleton() {}

    private static class Holder {
        private static ThreadSafeStaticInnerClassSingleton instance = new ThreadSafeStaticInnerClassSingleton();
    }

    public static ThreadSafeStaticInnerClassSingleton getInstance() {
        return Holder.instance;
    }
}

File: src/io/innofang/singleton/example/kotlin/ThreadSafeStaticInnerClassSingleton.kt
package io.innofang.singleton.example.kotlin

/**
 * Created by Inno Fang on 2017/8/12.
 */
class ThreadSafeStaticInnerClassSingleton {
    companion object {
        fun getInstance() = Holder.instance
    }

    private object Holder {
        val instance = ThreadSafeStaticInnerClassSingleton()
    }
}